skorch/_doctor.py

```diff
@@ -0,0 +1,936 @@
+                "'python -m pip install matplotlib'"
+            )
+            raise ImportError(msg) from exc
+
+        width = 12
+        height = 6
+
+        if figsize is None:
+            figsize = (width, nrows * height)
+
+        _, axes = plt.subplots(nrows, 1, figsize=figsize, squeeze=squeeze)
+        return axes
+
+    def plot_loss(self, ax=None, figsize=None, **kwargs):
+        """Plot the loss over each epoch.
+
+        Plots the training loss and, if present, the validation loss over time.
+
+        """
+        self.check_is_fitted()
+
+        ax = self._get_axes(ax, figsize=figsize, nrows=1, squeeze=True)
+        history = self.net.history
+        xvec = np.arange(len(history)) + 1
+        ax.plot(xvec, history[:, 'train_loss'], label='train', **kwargs)
+
+        if 'valid_loss' in history[0]:
+            ax.plot(xvec, history[:, 'valid_loss'], label='valid', **kwargs)
+
+        ax.set_xlabel('epoch')
+        ax.set_ylabel('loss')
+        ax.legend()
+        ax.set_title('loss over time')
+        return ax
+
+    def plot_activations(
+            self,
+            step=-1,
+            match_fn=None,
+            axes=None,
+            histtype='step',
+            lw=2,
+            bins=None,
+            density=True,
+            figsize=None,
+            **kwargs
+    ):
+        """Plot the distribution of activations produced by the layers
+
+        Parameters
+        ----------
+        step : int (default=-1)
+          Which training step to plot. By default, the last step (-1) is chosen.
+
+        match_fn : callable or None (default=None)
+          If not ``None``, this should be a callable/function that takes the
+          name of a layer as input and returns a bool as output, where ``False``
+          indicates that this layer should be excluded. Use this to filter only
+          specific layers you want to plot.
+
+        axes : np.ndarray of AxesSubplot or None (default=None)
+          By default, a new matplotlib plot is created. If you instead want to
+          plot onto an existing plot, pass it here. There should be one subplot
+          for each top level module (typically 2).
+
+        bins : np.ndarray or None (default=None)
+          Bins to use for the histogram. If left as ``None``, they are inferred
+          from the data.
+
+        **kwargs
+          You can override remaining plotting arguments like ``lw`` (line width)
+          or ``figsize`` (figure size).
+
+        Returns
+        -------
+        axes : np.andarray of AxesSubplot
+          The axes of the plot.
+
+        """
+        self.check_is_fitted()
+
+        # only use modules for which the values are not simply empty lists
+        module_names = [
+            key for key, val in self.activation_logs_.items() if any(l for l in val)
+        ]
+
+        axes = self._get_axes(axes, figsize=figsize, nrows=len(module_names))
+
+        for module_name, ax in zip(module_names, axes):
```

Comment:
Should the axes be flattened for cases where the user passes in axes that is "2d":

```
fig, axes = plt.subplots(3, 4)

doctor.plot_activations(..., axes=axes)
```